README
Module-Starter-Plugin-CGIApp

This is a plugin for Module::Starter that builds you a skeleton 
CGI::Application module with all the extra files needed to package it for 
CPAN. You can customize the output using HTML::Template.

INSTALLATION

If you have the Module::Signature module installed, you can verify the
integrity of this distribution by typing:

    cpansign -v

Module::Signature requires GnuPG which you can get from: 
http://www.gnupg.org/download/
A binary version for Windows is available from that page.
        
Then to install this module, run the following commands:

    perl Build.pl
    ./Build
    ./Build test
    ./Build install
    
On Windows you don't need to type the preliminary ./

Alternatively, you can do:

    perl Makefile.PL
    make
    make test
    make install

If you are using ActiveState perl on Windows and you don't have a make 
utility, you can download nmake from:
http://download.microsoft.com/download/vc15/patch/1.52/w95/en-us/nmake15.exe
(Replace make with nmake in the instructions above if you are using it.)

If you are using Strawberry perl on Windows, it comes with a make utility 
called dmake.  (Replace make with dmake in the instructions above if you are
using it.)

NOTE: At the moment, some of the tests will fail on Windows if the path to the 
directory you unpacked the distribution in has spaces in it.  Either skip the 
tests or move to another directory.

DEPENDENCIES

This module requires these other modules:

HTML-Template
Module-Starter

At build time you will need the above modules plus the following if you want
to run all the tests

Module-Signature
Test-Pod 1.22
Test-Pod-Coverage 1.04
Test-Perl-Critic (This test will only run if the env var TEST_AUTHOR has a 
                  true value)

They are all available on CPAN (http://www.cpan.org)


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc 
command.

    perldoc Module::Starter::Plugin::CGIApp

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/Module-Starter-Plugin-CGIApp

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Starter-Plugin-CGIApp

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/Module-Starter-Plugin-CGIApp

    CPAN Ratings:
        http://cpanratings.perl.org/d/Module-Starter-Plugin-CGIApp

COPYRIGHT AND LICENCE

Copyright (C) 2008,  Consolidated Braincells Inc. All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

$Id: README 82 2009-01-16 02:12:11Z jaldhar $


Changes
Revision history for Module-Starter-Plugin-CGIApp

0.42	Wed May 16 19:17:15 EDT 2012

        Internal:
        ---------

        The license of this module is "Same as Perl" which somewhere along
	    the line I thought meant GPL 2+/Artistic 2+ when actually it is 
	    GPL 1+/Artistic 1+.  This is now fixed in all the relevant places.

        In Created distributions:
        -------------------------

        The same license confusion occured here too.  Fixed.

0.41    Thu May 10 23:54:01 EDT 2012

        Internal:
        ---------

        Forgot to mention File::ShareDir as a dependency.

        While I'm at it, I normalized the format of copyright statements
        everywhere and updated the copyright year of this distro to 2012.

0.40    Thu May 10 22:58:23 EDT 2012

        Internal:
        ---------

        Breakage caused by newer versions of Module::Starter (RT#71734) has
        been fixed.

        Module has been updated to use the latest features of Module::Starter.

        Author tests have been moved into xt directory and brought up to
        current best practices.

        Templates have been moved into a share directory using File::ShareDir.

        Templates have been improved in various minor ways.

        In Created distributions:
        -------------------------

        Author tests are brought up to current best practices and now go into
        the xt directory so they don't have to be run by end users.

        Templates go into a shared directory using File::ShareDir instead
        of the old method of putting them in @INC

        Instead of pgp.mit.edu, t/00-signature.t will now use
        pool.sks-keyservers.net for key queries.

        MANIFEST.SKIP will also skip .git

0.30    Sat Mar  6 22:54:25 EST 2010
        cgiapp-starter and titanium-starter had got broken during the last
        release due to an errant local and were unable to find the default
        template directory. They should work correctly now.

        t/boilerplate.t in generated dists had got broken during the last
        set of updates.  It should work correctly now.  It will now also
        check that you do not have a boilerplate abstract in your buildscript.

        Previously there was support for a generated dist with multiple
        modules but it didn't really work.  Now it does.  This justifies
        a bump in the version number I think.

        The RT address in generated modules (for bug reporting) is based on
        the distro name instead of the module name now.

0.22    Thu Mar  4 01:52:09 EST 2010
        t/00-signature.t was wrong in the module and generated dists.
        Now it will only be carried out if the envvar TEST_SIGNATURE is set.

        Unneeded dependencies deleted from the build scripts in the module
        and generated dists.

        README in the module and generated dists updated to reflect these
        changes.

0.21    Wed Mar  3 15:28:02 EST 2010
        Fixed test failures on Windows (and possibly other platforms with
        disgusting perverted non-unixy line endings.)  Thanks csjewell,
        jdb, and samuraijack on #win32@irc.perl.org and almut, ambrus, and
        banesong on Perl Monks for their help in solving these problems.
        
        Upped the recommended perl version to 5.008.

0.20    Sun Feb 28 14:02:44 EST 2010
        t/test-app.t in generated dists now passes an object to $mech->app
        to match what is done is server.pl
        
        Tests in generated dists now follow Test::XT recommendations.
        
        If the license of a generated dist is 'perl', the license blurb in
        the module and README explicitly mentions GPL+Artistic instead of
        just 'same as perl'.  A LICENSE file is added to the dist where you
        can put the license terms.

        Hopefully the problem with Build mistakenly trying to run the 
        Makefile.PLs from templates is resolved.

        Revamped test suite should be more comprehensive than before.  In
        line with Test::XT recommendations, test relevant to the author
        only, will be skipped unless envvar RELEASE_TESTING=1.  See README.

0.10    Thu Jan 15 21:12:55 EST 2009
        Added a placeholder file in t/www in generated distributions as the
        empty directory was causing error messages.
        
        More build fixes for Windows.  I think I actually have got it right
        this time.  One problem remains.  Tests will fail if the directory 
        path to where you have unpacked the module has spaces anywhere in it.
        What's up with that?
        
        Ensured that MANIFEST in generated distributions always uses / as the
        directory seperator regardless of operating system.

        Added some entries to MANIFEST.SKIP in generated distributions to
        stop the signature test from failing under Windows.
        
        Thanks to Ron Savage and Cosimo Streppone from the cgiapp list and
        ikegami, cdarke, and andreas1234567 on Perl Monks for their help in 
        solving Windows issues.
        
0.08    Fri Jan  9 18:24:07 EST 2009
        This version just has changes to the build process which hopefully
        will make it work better under Windows.

0.07    Mon Jan  5 17:24:30 EST 2009
        No longer build depends on Test::Builder.
        
        By popular demand, templates use $self instead of $c.
        
        Distributions now include a directory, t/www for static files.
        server.pl has been updated to use it.
        
        Some fixes to tests so the module will build correctly on Windows
        thanks to David Scott <leapingfrog@yahoo.com>

        Some extra documentation for building on Windows added.

        HELP WANTED: despite my best efforts, I am not sure this module is 
        working correctly under Windows.  If you find problems, please
        submit a bug report, or even better, a fix.

0.05    Thu Sep 18 22:24:00 EDT 2008

t/common.pm
#
# $Id: common.pm 79 2009-01-15 19:38:11Z jaldhar $
#
package common;
use base qw( Exporter );
use warnings;
use strict;
use Cwd qw( cwd );
use English qw( -no_match_vars );
use File::Find qw();
use File::Spec;
use Test::More;

=head1 NAME

common - common functions and variables for this modules tests

=head1 VERSION

Version 1.2

=cut

our $VERSION = '1.2';

our @EXPORT = qw/ $cgiapp_starter $dir $root @expected_files &run_tests /;

our $dir  = File::Spec->catdir(cwd, 't');
our $root = File::Spec->catdir($dir, 'Foo');

our $cgiapp_starter;
if ($OSNAME =~ /win/i) {
    $cgiapp_starter = qq{ set MODULE_STARTER_DIR=$dir && cd $dir && $EXECUTABLE_NAME -Mblib ../script/cgiapp-starter --module=Foo --author="Jaldhar H. Vyas" --email=jaldhar\@braincells.com };
}
else {
    $cgiapp_starter = qq{ MODULE_STARTER_DIR=$dir ; export MODULE_STARTER_DIR ; cd $dir ; $EXECUTABLE_NAME -Mblib ../script/cgiapp-starter --module=Foo --author="Jaldhar H. Vyas" --email=jaldhar\@braincells.com };
}

our @expected_files = (
    'Foo',                   'Foo/lib',
    'Foo/lib/Foo.pm',        'Foo/lib/Foo',
    'Foo/lib/Foo/templates', 'Foo/lib/Foo/templates/runmode1.html',
    'Foo/t',                 'Foo/t/www/PUT.STATIC.CONTENT.HERE',
    'Foo/t/pod-coverage.t',  'Foo/t/pod.t',
    'Foo/t/test-app.t',      'Foo/t/01-load.t',
    'Foo/t/perl-critic.t',   'Foo/t/boilerplate.t',
    'Foo/t/00-signature.t',  'Foo/t/perlcriticrc',
    'Foo/Changes',           'Foo/README',
    'Foo/MANIFEST.SKIP',     'Foo/MANIFEST',
    'Foo/server.pl',         'Foo/t/www',
);

my $diag = '';

sub run_tests {
    my %got_files;

    # change / to local directory seperator in @expected_files
    map { $_ = File::Spec->catfile( split m{/}, $_ ) } @expected_files;
    
    # Now start by priming %got_files with the list of files we expect to 
    # exist.  The entry for each file will be given a value of -1
    foreach my $file (@expected_files) {
        $got_files{$file} = -1;
    }

    # Then add all the files that we actually found to %got_files.
    File::Find::find(
        {   no_chdir => 1,
            wanted  => sub {
                if ( -e $File::Find::name ) {
                    my $name = $File::Find::name;
                    $name = File::Spec->abs2rel($name, $dir);
                    # if the file we found is on the list of files expected 
                    # to exist, give its entry in %got_files a value of 1
                    # else the entry gets a value of 0.
                    $got_files{$name} = grep { $_ eq $name } @expected_files;
                    # furthermore if the file we found is on the list of files
                    # expected to exist and its length is non-zero, increase 
                    # the value of its entry in %got_files to 2.
                    if ( $got_files{$name} &&
                        ( -d $File::Find::name || -s $File::Find::name ) ) {
                        $got_files{$name}++;
                    }
                }
                return;
            }
        },
        $root,
    );

    # so to sum up, the value of the %got_files entry for a file can be...
    # -1  if the file is expected to exist but doesn't.
    #  0  if the file is not expected to exist but does.
    #  1  if the file is expected to exist and does but is zero-length.
    #  2  if the file is expected to exist and does and  is bigger than 0
    #     (which we assume means that it is good.)
    plan tests => ( scalar keys %got_files );

    foreach my $file ( keys %got_files ) {
        ok( diagnose_file($got_files{$file}), "$file$diag" );
    }
    
    return;
}

sub diagnose_file {
    my ($file) = @_;
    
    if ( $file == -1 ) {
        $diag = ' is missing';
        return;
    }
    elsif ( $file ==  0 ) {
        $diag = ' is extra';
        return;
    }
    elsif ( $file ==  1 ) {
        $diag = ' is zero-length';
        return;
    }
    $diag = '';
    return 1;
}

1;


lib/Module/Starter/Plugin/CGIApp/templates/titanium/Build.PL
#
# $Id: Build.PL 52 2009-01-06 03:22:31Z jaldhar $
#
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => '<tmpl_var main_module>',
    license             => '<tmpl_var license>',
    dist_author         => '<tmpl_var author> <<tmpl_var email>>',
    dist_version_from   => '<tmpl_var main_pm_file>',
    tmpl_files          => eval {
        my %tmpls; File::Find::find(
            {
                wanted => sub {
                    if ( -f $_ ) {
                        $tmpls{$File::Find::name} = $File::Find::name;
                    }
                },
            },
            '<tmpl_var templatedir>',
        );
        return \%tmpls;
    },
    build_requires => {
        'Titanium'                      => 0,
        'File::Find'                    => 0,      
        'HTML::Template'                => 0,
        'Pod::Coverage'                 => 0.18,
        'Pod::Usage'                    => 0,
        'Test::More'                    => 0,
        'Test::Pod'                     => 1.22,
        'Test::Pod::Coverage'           => 1.08,
        'Test::WWW::Mechanize::CGIApp'  => 0,
    },
    requires => {
        'Titanium'                      => 0,
        'File::Find'                    => 0,      
        'HTML::Template'                => 0,
    },
    add_to_cleanup      => [ '<tmpl_var distro>-*' ],
    create_makefile_pl  => 'traditional',
    sign                => 1,
);

$builder->add_build_element('tmpl');
$builder->create_build_script();



lib/Module/Starter/Plugin/CGIApp/templates/titanium/Module.pm
# $Id: Module.pm 52 2009-01-06 03:22:31Z jaldhar $

=head1 NAME

<tmpl_var module> - The great new <tmpl_var module> web application!

=head1 SYNOPSIS

    use <tmpl_var module>;
    my $app = <tmpl_var module>->new();
    $app->run();

=head1 ABSTRACT

A brief summary of what <tmpl_var module> does.

=cut

package <tmpl_var module>;

use warnings;
use strict;
use base 'Titanium';
use Carp qw( croak );

=head1 VERSION

This document describes <tmpl_var module> Version 0.01

=cut

our $VERSION = '0.01';

=head1 DESCRIPTION

Overview of functionality and purpose of
web application module <tmpl_var module>...

=head1 METHODS

=head2 SUBCLASSED METHODS

=head3 setup

Sets up the run mode dispatch table and the start, error, and default run modes.
If the template path is not set, sets it to a default value.

TODO: change all these values to ones more appropriate for your application.

=cut

sub setup {
    my ($self) = @_;

    $self->start_mode('runmode1');
    $self->error_mode('runmode1');
    $self->run_modes( [qw/ runmode1 /] );
    if ( !$self->tmpl_path ) {
        ( my $tp = '<tmpl_var module>' ) =~ s/::/\//gmx;
        ( $tp = $INC{"$tp.pm"} ) =~ s/.pm//mx;
        $self->tmpl_path("$tp/templates");
    }
    $self->run_modes( AUTOLOAD => 'runmode1' );
    return;
}

=pod

TODO: Other methods inherited from Titanium go here.

=head2 RUN MODES

=head3 runmode1

  * Purpose
  * Expected parameters
  * Function on success
  * Function on failure

TODO: Describe runmode1 here. Rename runmode1 to something more appropriate 
for your application.

=cut

sub runmode1 {
    my ($self) = @_;

    my $template = $self->load_tmpl;
    $template->param( message => 'Hello world!' );
    return $template->output;
}

=head2 OTHER METHODS

=head3 function1

TODO: Describe function1 here.  Rename function1 to something more appropriate
for your application.

=cut

sub function1 {
    my ($self) = @_;

    return 1;
}

=pod

TODO: Other methods in your public interface go here.

=cut

# TODO: Private methods go here. Start their names with an _ so they are skipped
# by Pod::Coverage.

=head1 BUGS AND LIMITATIONS

There are no known problems with this module.

Please report any bugs or feature requests to
C<bug-<tmpl_var rtname> at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=<tmpl_var distro>>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SEE ALSO

L<Titanium>

=head1 THANKS

List acknowledgements here or delete this section.

=head1 AUTHOR

<tmpl_var author>, C<< <<tmpl_var email_obfuscated>> >>

=head1 LICENSE AND COPYRIGHT

Copyright <tmpl_var year> <tmpl_var author>, all rights reserved.

<tmpl_var license_blurb>

=cut

1;    # End of <tmpl_var module>

__END__


lib/Module/Starter/Plugin/CGIApp/templates/titanium/Makefile.PL
#
# $Id: Makefile.PL 52 2009-01-06 03:22:31Z jaldhar $
#
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => '<tmpl_var main_module>',
    AUTHOR              => '<tmpl_var author> <<tmpl_var email>>',
    VERSION_FROM        => '<tmpl_var main_pm_file>',
    ABSTRACT_FROM       => '<tmpl_var main_pm_file>',
    PL_FILES            => {},
    PREREQ_PM => {
        'Titanium'                      => 0,
        'File::Find'                    => 0,
        'HTML::Template'                => 0,
        'Pod::Coverage'                 => 0.18,
        'Pod::Usage'                    => 0,
        'Test::More'                    => 0,
        'Test::Pod'                     => 1.22,
        'Test::Pod::Coverage'           => 1.08,
        'Test::WWW::Mechanize::CGIApp'  => 0,
    },
    SIGN => 1,
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => '<tmpl_var distro>-*' },
);


lib/Module/Starter/Plugin/CGIApp.pm
# $Id: CGIApp.pm 83 2009-01-16 02:15:34Z jaldhar $

=head1 NAME

Module::Starter::Plugin::CGIApp - template based module starter for CGI apps.

=head1 SYNOPSIS

    use Module::Starter qw(
        Module::Starter::Simple
        Module::Starter::Plugin::Template
        Module::Starter::Plugin::CGIApp
    );

    Module::Starter->create_distro(%args);

=head1 ABSTRACT

This is a plugin for L<Module::Starter> that builds you a skeleton 
L<CGI::Application> module with all the extra files needed to package it for 
CPAN. You can customize the output using L<HTML::Template>.

=cut

package Module::Starter::Plugin::CGIApp;

use warnings;
use strict;
use Carp qw( croak );
use English qw( -no_match_vars );
use ExtUtils::Command qw( mkpath );
use File::Basename;
use File::Spec ();
use Module::Starter::Simple;
use HTML::Template;

=head1 VERSION

Version 0.10

=cut

our $VERSION = '0.10';

=head1 DESCRIPTION

This module subclasses L<Module::Starter::Plugin::Template> which in turn 
subclasses L<Module::Starter::Simple>. This document only describes the methods
which are overriden from those modules or are new.

Only developers looking to extend this module need to read this. If you just 
want to use L<Module::Starter::Plugin::CGIApp>, read the docs for 
L<cgiapp-starter> or L<titanium-starter> instead.

=head1 METHODS

=head2 new ( %args )

This method calls the C<new> supermethod from 
L<Module::Starter::Plugin::Template> and then initializes the template store 
and renderer. (See C<templates> and C<renderer> below.)

=cut

sub new {
    my ( $class, @opts ) = @_;
    my $self = $class->SUPER::new(@opts);

    $self->{templates} = { $self->templates };
    $self->{renderer}  = $self->renderer;
    return bless $self => $class;
}

=head2 create_distro ( %args ) 

This method works as advertised in L<Module::Starter>.

=cut

sub create_distro {
    my ( $class, @opts ) = @_;

    my $self = $class->new(@opts);

    my @modules = map { split /,/mx } @{ $self->{modules} };

    if ( !@modules ) {
        croak "No modules specified.\n";
    }
    for (@modules) {
        if ( !/\A[a-z_]\w*(?:::[\w]+)*\Z/imx ) {
            croak "Invalid module name: $_";
        }
    }

    if ( !$self->{author} ) {
        croak "Must specify an author\n";
    }
    if ( !$self->{email} ) {
        croak "Must specify an email address\n";
    }
    ( $self->{email_obfuscated} = $self->{email} ) =~ s/@/ at /mx;

    $self->{license} ||= 'perl';

    $self->{main_module} = $self->{modules}->[0];
    if ( !$self->{distro} ) {
        $self->{distro} = $self->{main_module};
        $self->{distro} =~ s/::/-/gmx;
    }

    $self->{basedir} = $self->{dir} || $self->{distro};
    $self->create_basedir;

    my @distroparts = split /-/mx, $self->{distro};
    $self->{templatedir}
        = File::Spec->catdir( 'lib', @distroparts, 'templates' );

    my @files;
    push @files, $self->create_modules(@modules);

    push @files, $self->create_t(@modules);
    push @files, $self->create_tmpl();
    my %build_results = $self->create_build();
    push @files, @{ $build_results{files} };

    push @files, $self->create_Changes;
    push @files, $self->create_README( $build_results{instructions} );
    push @files, $self->create_MANIFEST_SKIP;
    push @files, $self->create_perlcriticrc;
    push @files, $self->create_server_pl;
    push @files, 'MANIFEST';
    $self->create_MANIFEST( grep { $_ ne 't/boilerplate.t' } @files );

    return;
}

=head2 create_MANIFEST_SKIP( )

This method creates a C<MANIFEST.SKIP> file in the distribution's directory so 
that unneeded files can be skipped from inclusion in the distribution.

=cut

sub create_MANIFEST_SKIP {
    my $self = shift;

    my $fname = File::Spec->catfile( $self->{basedir}, 'MANIFEST.SKIP' );
    $self->create_file( $fname, $self->MANIFEST_SKIP_guts() );
    $self->progress("Created $fname");

    return 'MANIFEST.SKIP';
}

=head2 create_perlcriticrc( )

This method creates a C<perlcriticrc> in the distribution's test directory so 
that the behavior of C<perl-critic.t> can be modified.

=cut

sub create_perlcriticrc {
    my $self = shift;

    my @dirparts = ( $self->{basedir}, 't' );
    my $tdir = File::Spec->catdir(@dirparts);
    if ( not -d $tdir ) {
        local @ARGV = $tdir;
        mkpath();
        $self->progress("Created $tdir");
    }

    my $fname = File::Spec->catfile( @dirparts, 'perlcriticrc' );
    $self->create_file( $fname, $self->perlcriticrc_guts() );
    $self->progress("Created $fname");

    return 't/perlcriticrc';
}

=head2 create_server_pl( )

This method creates C<server.pl> in the distribution's root directory.

=cut

sub create_server_pl {
    my $self = shift;

    my $fname = File::Spec->catfile( $self->{basedir}, 'server.pl' );
    $self->create_file( $fname, $self->server_pl_guts() );
    $self->progress("Created $fname");

    return 'server.pl';
}

=head2 create_t( @modules )

This method creates a bunch of *.t files.  I<@modules> is a list of all modules
in the distribution.

=cut

sub create_t {
    my ( $self, @modules ) = shift;

    my %t_files = $self->t_guts(@modules);

    my @files = map { $self->_create_t( $_, $t_files{$_} ) } keys %t_files;

    # This next part is for the static files dir t/www
    my @dirparts = ( $self->{basedir}, 't', 'www' );
    my $twdir = File::Spec->catdir(@dirparts);
    if ( not -d $twdir ) {
        local @ARGV = $twdir;
        mkpath();
        $self->progress("Created $twdir");
    }
    my $placeholder
        = File::Spec->catfile( @dirparts, 'PUT.STATIC.CONTENT.HERE' );
    $self->create_file( $placeholder, q{ } );
    $self->progress("Created $placeholder");
    push @files, 't/www/PUT.STATIC.CONTENT.HERE';

    return @files;
}

=head2 create_tmpl( )

This method takes all the template files ending in .html (representing 
L<HTML::Template>'s and installs them into a directory under the distro tree.  
For instance if the distro was called C<Foo-Bar>, the templates would be 
installed in C<lib/Foo/Bar/templates>.

Note the files will just be copied over not rendered.

=cut

sub create_tmpl {
    my $self = shift;

    return $self->tmpl_guts();
}

=head2 render( $template, \%options )

This method is subclassed from L<Module::Starter::Plugin::Template>.

It is given an L<HTML::Template> and options and returns the resulting document.

Data in the C<Module::Starter> object which represents a reference to an array 
@foo is transformed into an array of hashes with one key called 
C<$foo_item> in order to make it usable in an L<HTML::Template> C<TMPL_LOOP>.
For example:

    $data = ['a'. 'b', 'c'];

would become:

    $data = [
        { data_item => 'a' },
        { data_item => 'b' },
        { data_item => 'c' },
    ];
    
so that in the template you could say:

    <tmpl_loop data>
        <tmpl_var data_item>
    </tmpl_loop>
    
=cut

sub render {
    my ( $self, $template, $options ) = @_;

    # we need a local copy of $options otherwise we get recursion in loops
    # because of [1]
    my %opts = %{$options};

    $opts{nummodules}    = scalar @{ $self->{modules} };
    $opts{year}          = $self->_thisyear();
    $opts{license_blurb} = $self->_license_blurb();
    $opts{datetime}      = scalar localtime;

    foreach my $key ( keys %{$self} ) {
        next if defined $opts{$key};
        $opts{$key} = $self->{$key};
    }

    # [1] HTML::Templates wants loops to be arrays of hashes not plain arrays
    foreach my $key ( keys %opts ) {
        if ( ref $opts{$key} eq 'ARRAY' ) {
            my @temp = ();
            for my $option ( @{ $opts{$key} } ) {
                push @temp, { "${key}_item" => $option };
            }
            $opts{$key} = [@temp];
        }
    }
    my $t = HTML::Template->new(
        die_on_bad_params => 0,
        scalarref         => \$template,
    ) or croak "Can't create template $template";
    $t->param( \%opts );
    return $t->output;
}

=head2 renderer ( )

This method is subclassed from L<Module::Starter::Plugin::Template> but
doesn't do anything as the actual template is created by C<render> in this 
implementation.

=cut

sub renderer {
    my ($self) = @_;
    return;
}

=head2 templates ( )

This method is subclassed from L<Module::Starter::Plugin::Template>.

It reads in the template files and populates the object's templates 
attribute. The module template directory is found by checking the 
C<MODULE_TEMPLATE_DIR> environment variable and then the config option 
C<template_dir>.

=cut

sub templates {
    my ($self) = @_;
    my %template;

    my $template_dir = ( $ENV{MODULE_TEMPLATE_DIR} || $self->{template_dir} )
        or croak 'template dir not defined';
    if ( !-d $template_dir ) {
        croak "template dir does not exist: $template_dir";
    }

    foreach ( glob "$template_dir/*" ) {
        my $basename = basename $_;
        next if ( not -f $_ ) or ( $basename =~ /^\./mx );
        open my $template_file, '<', $_
            or croak "couldn't open template: $_";
        $template{$basename} = do {
            local $RS = undef;
            <$template_file>;
        };
        close $template_file or croak "couldn't close template: $_";
    }

    return %template;
}

=head2 Changes_guts

Implements the creation of a C<Changes> file.

=cut

sub Changes_guts {    ## no critic (ProhibitMixedCaseSubs)
    my $self = shift;
    my %options;

    my $template = $self->{templates}{Changes};
    return $self->render( $template, \%options );
}

=head2 MANIFEST_SKIP_guts

Implements the creation of a C<MANIFEST.SKIP> file.

=cut

sub MANIFEST_SKIP_guts {
    my $self = shift;
    my %options;

    my $template = $self->{templates}{'MANIFEST.SKIP'};
    return $self->render( $template, \%options );
}

=head2 perlcriticrc_guts

Implements the creation of a C<perlcriticrc> file.
 
=cut

sub perlcriticrc_guts {
    my $self = shift;
    my %options;

    my $template = $self->{templates}{perlcriticrc};
    return $self->render( $template, \%options );
}

=head2 server_pl_guts

Implements the creation of a C<server.pl> file.

=cut

sub server_pl_guts {
    my $self = shift;
    my %options;
    $options{main_module} = $self->{main_module};

    my $template = $self->{templates}{'server.pl'};
    return $self->render( $template, \%options );
}

=head2 t_guts

Implements the creation of test files.

=cut

sub t_guts {
    my ( $self, @opts ) = @_;
    my %options;
    $options{modules}     = [@opts];
    $options{modulenames} = [];
    foreach ( @{ $options{modules} } ) {
        push @{ $options{module_pm_files} }, $self->_module_to_pm_file($_);
    }

    my %t_files;

    foreach ( grep {/\.t$/mx} keys %{ $self->{templates} } ) {
        my $template = $self->{templates}{$_};
        $t_files{$_} = $self->render( $template, \%options );
    }

    return %t_files;
}

=head2 tmpl_guts

Implements the creation of template files.

=cut

sub tmpl_guts {
    my ($self) = @_;
    my %options;    # unused in this function.

    # we need the directory seperator to be / regardless of OS
    my $reldir = join q{/}, File::Spec->splitdir( $self->{templatedir} );
    my @dirparts = ( $self->{basedir}, $self->{templatedir} );
    my $tdir = File::Spec->catdir(@dirparts);
    if ( not -d $tdir ) {
        local @ARGV = $tdir;
        mkpath();
        $self->progress("Created $tdir");
    }

    my @t_files;
    foreach my $filename ( grep {/\.html$/mx} keys %{ $self->{templates} } ) {
        my $template = $self->{templates}{$filename};
        my $fname = File::Spec->catfile( @dirparts, $filename );
        $self->create_file( $fname, $template );
        $self->progress("Created $fname");
        push @t_files, "$reldir/$filename";
    }

    return @t_files;
}

=head1 BUGS

Please report any bugs or feature requests to 
C<bug-module-starter-plugin-cgiapp at rt.cpan.org>, or through the web 
interface at L<http://rt.cpan.org>. I will be notified, and then you'll 
automatically be notified of progress on your bug as I make changes.

=head1 AUTHOR

Jaldhar H. Vyas, E<lt>jaldhar at braincells.comE<gt>

=head1 COPYRIGHT

Copyright (C) 2008,  Consolidated Braincells Inc. All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 SEE ALSO

L<cgiapp-starter>, L<titanium-starter>, L<Module::Starter>, 
L<Module::Starter::Simple>, L<Module::Starter::Plugin::Template>. 
L<CGI::Application>, L<Titanium>, L<HTML::Template>

=cut

1;

